File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

public class ProteinTranslation
{
    private static Dictionary<string, string> codonProteinMap = new Dictionary<string, string>
    {
        ["AUG"] = "Methionine",
        ["UUU"] = "Phenylalanine", ["UUC"] = "Phenylalanine",
        ["UUA"] = "Leucine", ["UUG"] = "Leucine",
        ["UCU"] = "Serine", ["UCC"] = "Serine", ["UCA"] = "Serine", ["UCG"] = "Serine",
        ["UAU"] = "Tyrosine", ["UAC"] = "Tyrosine",
        ["UGU"] = "Cysteine", ["UGC"] = "Cysteine",
        ["UGG"] = "Tryptophan",
        ["UAA"] = "STOP", ["UAG"] = "STOP", ["UGA"] = "STOP"
    };

    public static string[] Proteins(string rna)
    {
        var codons = Regex.Matches(rna, "[A-Z]{3}")
                          .Cast<Match>()
                          .Select(e => e.Value)
                          .ToArray();

        List<string> proteins = new List<string>(codons.Length);
        foreach (var codon in codons)
        {
            if (!codonProteinMap.ContainsKey(codon))
            {
                throw new Exception($"Invalid codon: {codon}");
            }

            if (codonProteinMap[codon] == "STOP")
            {
                break;
            }

            proteins.Add(codonProteinMap[codon]);
        }
            
        return proteins.ToArray();
    }
}

File: javascript/protein-translation/protein-translation.js
const codonProteinMap = {
    "AUG": "Methionine",
    "UUU": "Phenylalanine",
    "UUC": "Phenylalanine",
    "UUA": "Leucine",
    "UUG": "Leucine",
    "UCU": "Serine",
    "UCC": "Serine",
    "UCA": "Serine",
    "UCG": "Serine",
    "UAU": "Tyrosine",
    "UAC": "Tyrosine",
    "UGU": "Cysteine",
    "UGC": "Cysteine",
    "UGG": "Tryptophan",
    "UAA": "STOP",
    "UAG": "STOP",
    "UGA": "STOP"
}

var translate = function(rna = "") {
    let polypeptide = [];
    let codons = rna.match(/[A-Z]{3}/g) || [];

    for (let i = 0; i < codons.length; i++) {
        if (codonProteinMap[codons[i]] === undefined) {
            throw new Error("Invalid codon");
        }
        else if (codonProteinMap[codons[i]] === "STOP") {
            break;
        }
        else {
            polypeptide.push(codonProteinMap[codons[i]]);
        }
    }

    return polypeptide;
};

module.exports = translate;
